controllers/kubeadmbootstrapconfig_controller_test.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	kubeadmv1alpha1 "sigs.k8s.io/cluster-api-bootstrap-provider-kubeadm/api/v1alpha1"
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type log struct{}
+
+func (l *log) Info(msg string, keysAndValues ...interface{})             {}
+func (l *log) Enabled() bool                                             { return false }
+func (l *log) Error(err error, msg string, keysAndValues ...interface{}) {}
+func (l *log) V(level int) logr.InfoLogger                               { return l }
+func (l *log) WithValues(keysAndValues ...interface{}) logr.Logger       { return l }
+func (l *log) WithName(name string) logr.Logger                          { return l }
```

Comment:
do all unit tests in different files have to define a mock logger like so?

Reply:
Sometimes nil will suffice.

That being said, there is nothing stopping us from having a `bootstraptest` package that contains shared mocks _if the mocks are sufficiently dumb/reusable_. We do this in cluster-api-provider-aws with *generated* mocks, but shared mocks do not have to be generated.

Keeping mocks as thin and dumb as possible and not trying to reuse the same one for many kinds of functionality is crucial to maintaining clarity. Ideally mocks should contain no logic, but in practice they will need a tiny bit of logic.